### PR TITLE
[StableHash] Implement stable global name for the hash computation

### PR DESCRIPTION
**Pull request by kyulee-com**
LLVM often extends global names by adding suffixes to distinguish unique identities. However, these suffixes are not always stable across different runs and build environments. To address this issue, I implemented `get_stable_name` to ignore such suffixes and obtain the original name. This approach is not new, as PGO or Bolt already handle this issue similarly. Using the stable name obtained from `get_stable_name`, I implemented `stable_hash_name` while utilizing the same underlying `xxh3_64bit` algorithm as before.

### CODE REVIEW

File: llvm/include/llvm/ADT/StableHashing.h [hunk 1]
```diff
@@ -50,6 +50,22 @@ inline stable_hash stable_hash_combine(stable_hash A, stable_hash B,
   return stable_hash_combine(Hashes);
 }
 
+// Removes suffixes introduced by LLVM from the name to enhance stability and
+// maintain closeness to the original name across different builds.
+inline StringRef get_stable_name(StringRef Name) {
+  auto [P1, S1] = Name.rsplit(".llvm.");
+  auto [P2, S2] = P1.rsplit(".__uniq.");
+  return P2;
+}
```
Comment (ellishg) [inline llvm/include/llvm/ADT/StableHashing.h:59]: Similar code is found in a few more place.

https://github.com/search?q=%5C%22.__uniq.%5C%22+repo%3Allvm%2Fllvm-project+path%3A*.cpp&type=code&ref=advsearch

I wonder if it makes sense to put this code somewhere so that the above functions can be replaced with it. That might be better off in a separate PR, though.

File: llvm/include/llvm/ADT/StableHashing.h [hunk 2]
```diff
@@ -50,6 +50,22 @@ inline stable_hash stable_hash_combine(stable_hash A, stable_hash B,
...
+
+// Generates a consistent hash value for a given input name across different
+// program executions and environments. This function first converts the input
+// name into a stable form using the `get_stable_name` function, and then
+// computes a hash of this stable name.
```
Comment (ellishg) [inline llvm/include/llvm/ADT/StableHashing.h:64]: I guess `foo.llvm.1234` would have the same hash as `foo.llvm.5678`. It might be worth mentioning that hash collisions are not uncommon.

File: llvm/lib/CodeGen/MachineStableHash.cpp [hunk 3]
```diff
@@ -95,13 +95,21 @@ stable_hash llvm::stableHashValue(const MachineOperand &MO) {
   case MachineOperand::MO_Metadata:
     StableHashBailingMetadataUnsupported++;
     return 0;
-  case MachineOperand::MO_GlobalAddress:
-    StableHashBailingGlobalAddress++;
-    return 0;
+  case MachineOperand::MO_GlobalAddress: {
+    const GlobalValue *GV = MO.getGlobal();
+    if (!GV->hasName()) {
+      StableHashBailingGlobalAddress++;
```
Comment (ellishg) [inline llvm/lib/CodeGen/MachineStableHash.cpp:101]:
```suggestion
      ++StableHashBailingGlobalAddress;
```
Prefer preincrement, but you might also want to change the rest of the stats variables.
https://llvm.org/docs/CodingStandards.html#prefer-preincrement

File: llvm/unittests/MIR/MachineStableHashTest.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  MachineModuleInfo MMI(TM.get());
+  M = parseMIR(*TM, MIRString, MMI);
+  ASSERT_TRUE(M);
+  auto *MF1 = MMI.getMachineFunction(*M->getFunction("f1"));
+  auto *MF2 = MMI.getMachineFunction(*M->getFunction("f2"));
+  auto *MF3 = MMI.getMachineFunction(*M->getFunction("f3"));
+  auto *MF4 = MMI.getMachineFunction(*M->getFunction("f4"));
+
+  // Expect the suffix, `.llvm.{number}` to be ignored.
+  EXPECT_EQ(stableHashValue(*MF1), stableHashValue(*MF2));
```
Comment (ellishg) [inline llvm/unittests/MIR/MachineStableHashTest.cpp:138]:
```suggestion
  EXPECT_EQ(stableHashValue(*MF1), stableHashValue(*MF2)) << "Expect the suffix, `.llvm.{number}` to be ignored.";
```
It would be nice to have the explanation in the error message if the test breaks.

File: llvm/include/llvm/ADT/StableHashing.h (hunk 1 above)
Comment (kyulee-com) [reply llvm/include/llvm/ADT/StableHashing.h:59]: Yes, the code segments appear similar, but they could be implied slightly differently depending on the client. Specifically, in this instance, I am planning to further extend the functionality for the upcoming global outlining feature -- https://github.com/llvm/llvm-project/pull/90074. I intend to introduce an additional suffix `.content.{number}` to the outlined function names, which will enable us to deduce the identities of target functions even when their names vary -- in that case, use the content hash from the suffix, rather than ignoring it . So, I think the usage here is primarily focused on accurately identifying the stable target based on the name.

File: llvm/unittests/MIR/MachineStableHashTest.cpp (hunk 4 above)
Comment (kyulee-com) [reply llvm/unittests/MIR/MachineStableHashTest.cpp]: Thanks for the suggestion. Added the expected output message.

File: llvm/lib/CodeGen/MachineStableHash.cpp (hunk 3 above)
Comment (kyulee-com) [reply llvm/lib/CodeGen/MachineStableHash.cpp]: Updated the increments of the stats variables.

File: llvm/include/llvm/ADT/StableHashing.h (hunk 2 above)
Comment (kyulee-com) [reply llvm/include/llvm/ADT/StableHashing.h]: Added comments.

**Review by ellishg — APPROVED** [review]
LGTM

**Comment by kyulee-com** [comment]
Thank you for the prompt review! I will address any further issues in the next PR if they arise.
